src/gt4py/next/common.py

```diff
@@ -61,21 +65,167 @@ def __str__(self):
         return f'Dimension(value="{self.value}", kind={self.kind})'
 
 
-class DType:
-    ...
+DomainLike = Union[Sequence[Dimension], Dimension, str]
 
 
-class Field(Generic[DimsT, DT]):
-    ...
+@dataclasses.dataclass(frozen=True)
+class UnitRange(Sequence, Set):
+    """Range from `start` to `stop` with step size one."""
 
+    start: int
+    stop: int
 
-@dataclass(frozen=True)
+    def __len__(self) -> int:
+        return max(0, self.stop - self.start)
+
+    def __repr__(self) -> str:
+        return f"UnitRange({self.start}, {self.stop})"
+
+    @overload
+    def __getitem__(self, index: int) -> int:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> list[int]:
+        ...
+
+    def __getitem__(self, index: int | slice) -> int | list[int]:
+        if isinstance(index, slice):
+            start, stop, step = index.indices(len(self))
+            return [self[i] for i in range(start, stop, step)]
```

Comment:
this should probably return UnitRange.
Is requesting a larger slice an implicit intersect or an error?